iroh-net/src/discovery/pkarr_publish.rs

```diff
@@ -0,0 +1,116 @@
+//! A discovery service which publishes node information to a [Pkarr] relay.
+//!
+//! This service only implements the [`Discovery::publish`] method and does not provide discovery.
+//! It encodes the node information into a DNS packet in the format resolvable by the
+//! [`super::dns::DnsDiscovery`], which means a single _iroh_node TXT record, under the z32 encoded
+//! node id as origin domain.
+//!
+//! [pkarr]: https://pkarr.org
+
+// TODO: Decide what to do with this module once publishing over relays land. Either remove, or
+// leave in the repo but do not enable it by default in the iroh node.
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use parking_lot::RwLock;
+use pkarr::SignedPacket;
+use tracing::warn;
+use url::Url;
+
+use crate::{discovery::Discovery, dns::node_info::NodeInfo, key::SecretKey, AddrInfo};
+
+/// The n0 testing pkarr relay
+pub const N0_TESTDNS_PKARR_RELAY: &str = "https://testdns.iroh.link/pkarr";
+
+/// Default TTL for the _iroh_node TXT record in the pkarr signed packet
+const DEFAULT_PKARR_TTL: u32 = 30;
+
+/// Publish node info to a pkarr relay.
+#[derive(derive_more::Debug, Clone)]
+pub struct Publisher {
+    #[debug("SecretKey")]
+    secret_key: SecretKey,
+    #[debug("PkarrClient")]
+    pkarr_client: PkarrRelayClient,
+    last_published: Arc<RwLock<Option<NodeInfo>>>,
+    ttl: u32,
+}
+
+impl Publisher {
+    /// Create a new config with a secret key and a pkarr relay URL.
+    pub fn new(secret_key: SecretKey, pkarr_relay: Url) -> Self {
+        let pkarr_client = PkarrRelayClient::new(pkarr_relay);
+        Self {
+            secret_key,
+            pkarr_client,
+            ttl: DEFAULT_PKARR_TTL,
+            last_published: Default::default(),
+        }
+    }
+
+    /// Create a config that publishes to the n0 testdns server.
+    pub fn n0_testdns(secret_key: SecretKey) -> Self {
+        let pkarr_relay: Url = N0_TESTDNS_PKARR_RELAY.parse().expect("url is valid");
+        Self::new(secret_key, pkarr_relay)
+    }
+
+    /// Set the TTL for pkarr packets, in seconds.
+    ///
+    /// Default value is 30 seconds.
+    pub fn set_ttl(&mut self, ttl: u32) {
+        self.ttl = ttl;
+    }
+
+    /// Publish [`AddrInfo`] about this node to a pkarr relay.
+    pub async fn publish_addr_info(&self, info: &AddrInfo) -> Result<()> {
```

Comment:
This does not do republishing. If I remember correctly, the pkarr crate also does not do it. So this will not republish. So the info in the DHT will vanish after a few days.

In iroh-pkarr-node-discovery I spawn a task that republishes. See https://github.com/n0-computer/iroh-examples/blob/main/iroh-pkarr-node-discovery/src/lib.rs#L109

Or is the publishing triggered in regular intervals by the magicsock?

Reply:
By the way: in the iroh-pkarr-node-discovery crate I just have a small initial delay before publishing in the task, that way if the addr changes frequently before settling down the transient addrs are not published.

Reply:
This only publishes to a pkarr relay over http, not to mainline, so I don't think it needs republishing. If we were to change that to publish to mainline directly we would need republishing, true.

Reply:
So the idea is that the pkarr relay will continue publishing this information even if it is very old? Or will it somehow detect that the node has not been seen for a while and discard the info?

E.g. you start a test that creates a large number of random node ids. We might do this, but even if we don't, _somebody_ will. So then you are spamming the DHT with thousands of publish requests that are no longer relevant, and risk being rate limited by the DHT nodes.

I guess the derper has all the info it needs to check if info is still relevant.

But still I somehow like the idea of very low frequency republishing to make things self-healing.

E.g. you publish, the publish gets lost because the derper crashed. Now you are in a state where you think your info is publicly known, but it won't be until your network changes (which might be never for a static node).

In general, whenever there is some system that requires reliable delivery of every single message I think it becomes much more fragile compared to if you have a way to avoid this. E.g. a progress update stream that requires the remote side to sum the deltas compared to a progress stream that occasionally sends the full current value to reset the state.

I would also like it if we would retain the ability to publish directly to the DHT via pkarr, at least as an option. So this is not a regression compared to the stuff I hacked with @arnazeh.

Reply:
The derper would only once publish the info it receives onto a pkarr relay. Our pkarr relay/DNS server stores and serves everything, forever (or until replaced by update). So also for publish-via-relay: no republishing yet because no mainline DHT integration, thus not needed.

I do agree though that there's no downside and some upsides to eventual republishing. It will increase load a bit on our DNS server/pkarr relay, but well. Maybe once every 5 minutes or so? Not sure.

> I would also like it if we would retain the ability to publish directly to the DHT via pkarr,
I am supportive of this! But would like to move to followups and not do in this PR.

Reply:
See #2151 for a change to do republishing. I am quite unsure though what the republish interval should be. WDYT? Let's take the discussion to 2151 though.